File: 0976-minimum-area-rectangle/0976-minimum-area-rectangle.js
/**
 * @param {number[][]} points
 * @return {number}
 */
var minAreaRect = function(points) {
    let xAxis = {}; //x좌표가 같은것들
    let yAxis = {}; //y좌표가 같은 것들 

    let minArea = Infinity;

    for(let [x,y] of points){
        if(!xAxis[x]) xAxis[x] = [y];
        else xAxis[x].push(y);

        if(!yAxis[y]) yAxis[y] = [x];
        else yAxis[y].push(x);
    }

    //[x,y] 가 같은거 순회해서 map에서 찾고, 대각선 x좌표가 같은거 찾기 , y 좌표가 같은거 찾기 

    //네점이 주어지고 네 점이 직사각형이 구해야함 -> 대각선의 점과 한 점이 주어지면 직사각형이 그려짐 
    //이중 for문을 그리면됨 
    for(let i = 0 ; i < points.length ;i++){
        let [px0,py0] = points[i]
        for(let j = i + 1 ; j <points.length ;j++){
            //두 점을 가리키는 포인터 
            //[1,1] 과 [1,3]이라면, 두
            //x에서 찾기 
            let [px1,py1] = points[j];

            curArea = Math.abs(px0 - px1) * Math.abs(py0 - py1)
            if (curArea ==0){
                continue;
            }
            
            //만약에 px0 의 py1가 xAxis에 없다면, 
            if(!xAxis[px0].includes(py1)){
                continue;
            }         

            if(!xAxis[px1].includes(py0)){
                continue;
            }

            minArea = Math.min(minArea, curArea);

        }
    }
    if(minArea == Infinity){
        minArea =0
    }
    
    //만약에 1,1 -> 1,3 , 3,1 
    return minArea
};


//대각선의 길이가 같아야함

//x, y 좌표로 좌표를 줌 

// 만약에 직사각형이 아니라면, return 0;

// 직사각형중 가장 크기가 작은거

//네 점을 구해야함 대각선의 길이가 같고 x끼리 같은거 y끼리 같은거 구해야함

File: 0976-minimum-area-rectangle/0976-minimum-area-rectangle.go
import (
    "fmt"
    "math"
)

func minAreaRect(points [][]int) int {

    pointSet := make(map[[2]int]bool) //점 존재 여부를 확인하기 위한 맵 
    minArea := math.MaxInt32

    for _, point := range points {
        x, y := point[0], point[1]
        pointSet[[2]int{x,y}] = true;
    }



    for i := 0 ; i < len(points) -1 ; i++ {
        x0, y0 := points[i][0], points[i][1]; 
        for j := i + 1 ; j < len(points) -1 ; j++ {
            x1, y1 := points[j][0], points[j][1];

            if x0 != x1 && y0 != y1 {
                if pointSet[[2]int{x0, y1}] && pointSet[[2]int{x1, y0}] {
                    area := int(math.Abs(float64(x1 - x0)) * math.Abs(float64(y1 - y0)))
                    if area < minArea {
                        minArea = area
                    }
                }
            }

            
        }
 
    }


    if minArea == math.MaxInt32 {
        return 0
    }


    return minArea
}

//두 점을 구하고, 두점의 직 사각형이 가장 작은 사각형 구하기